src/runtime/webgpu.cpp

```diff
@@ -0,0 +1,1146 @@
+            wait();
+        }
+    }
+
+    // Wait for all error callbacks in this scope to fire.
+    // Returns the error code (or success).
+    int wait() {
+        if (callbacks_remaining == 0) {
+            error(user_context) << "no outstanding error scopes\n";
+            return halide_error_code_internal_error;
+        }
+
+        error_code = halide_error_code_success;
+        wgpuDevicePopErrorScope(device, error_callback, this);
+        wgpuDevicePopErrorScope(device, error_callback, this);
+
+        // Wait for the error callbacks to fire.
+        while (__sync_fetch_and_or(&callbacks_remaining, 0) > 0) {
+            wgpuDeviceTick(device);
+        }
+
+        return error_code;
+    }
+
+private:
+    void *user_context;
+    WGPUDevice device;
+
+    // The error code reported by the callback functions.
+    volatile int error_code;
+
+    // Used to track outstanding error callbacks.
+    volatile int callbacks_remaining = 0;
+
+    // The error callback function.
+    // Logs any errors, and decrements the remaining callback count.
+    static void error_callback(WGPUErrorType type,
+                               char const *message,
+                               void *userdata) {
+        ErrorScope *context = (ErrorScope *)userdata;
+        switch (type) {
+        case WGPUErrorType_NoError:
+            // Do not overwrite the error_code to avoid masking earlier errors.
+            break;
+        case WGPUErrorType_Validation:
+            error(context->user_context) << "WGPU: validation error: "
+                                         << message << "\n";
+            context->error_code = halide_error_code_generic_error;
+            break;
+        case WGPUErrorType_OutOfMemory:
+            error(context->user_context) << "WGPU: out-of-memory error: "
+                                         << message << "\n";
+            context->error_code = halide_error_code_out_of_memory;
+            break;
+        default:
+            error(context->user_context) << "WGPU: unknown error (" << type
+                                         << "): " << message << "\n";
+            context->error_code = halide_error_code_generic_error;
+        }
+
+        __sync_fetch_and_add(&context->callbacks_remaining, -1);
+    }
+};
+
+// WgpuBufferHandle represents a device buffer with an offset.
+struct WgpuBufferHandle {
+    uint64_t offset;
+    WGPUBuffer buffer;
+};
+
+// A cache for compiled WGSL shader modules.
+WEAK Halide::Internal::GPUCompilationCache<WGPUDevice, WGPUShaderModule>
+    shader_cache;
+
+namespace {
+
+halide_error_code_t init_error_code = halide_error_code_success;
+
+void device_lost_callback(WGPUDeviceLostReason reason,
+                          char const *message,
+                          void *user_context) {
+    error(user_context) << "WGPU device lost (" << reason << "): "
+                        << message << "\n";
```

Comment:
Error handling in the Halide runtime currently requires that you call `halide_error()` (or use `error()`, which does that for you), **and** return a nonzero value to the (eventual) caller -- in most production code, `halide_error()` is replaced to simply log the message (and not call `abort()`, etc), so this error would get dropped on the floor in those cases.

This appears to be the only place here where we don't return a nonzero value (eventually)... but it's not clear to me how and where we would need to do that. (Set another global that is checked... somewhere?)

Reply:
I've added a global flag to capture device lost, which is checked on the next `acquire_context`.